packages/graphile-build-pg/src/plugins/PgTypesPlugin.js

```diff
@@ -997,31 +1000,58 @@ end`;
           type.type === "d" &&
           type.domainBaseTypeId
         ) {
-          const baseType = getGqlTypeByTypeIdAndModifier(
-            type.domainBaseTypeId,
-            typeModifier
-          );
-          const baseInputType = getGqlInputTypeByTypeIdAndModifier(
-            type.domainBaseTypeId,
-            typeModifier
-          );
-          // Hack stolen from: https://github.com/graphile/postgraphile/blob/ade728ed8f8e3ecdc5fdad7d770c67aa573578eb/src/graphql/schema/type/aliasGqlType.ts#L16
-          gqlTypeByTypeIdAndModifier[type.id][typeModifierKey] = Object.assign(
-            Object.create(baseType),
-            {
-              name: inflection.domainType(type),
-              description: type.description,
-            }
-          );
-          if (baseInputType && baseInputType !== baseType) {
+          // might be used as an enum alias: #1500
+          if (
+            type.name.endsWith("_enum_domain") ||
+            (type.tags && type.tags.enum)
+          ) {
+            const underlyingEnumTypeName = type.name.endsWith("_enum_domain")
+              ? type.name.replace("_enum_domain", "")
+              : type.tags.enum;
+
+            const baseTypeId = `FAKE_ENUM_${type.namespaceName}_${underlyingEnumTypeName}`;
+
+            const baseType = getGqlTypeByTypeIdAndModifier(
+              baseTypeId,
+              typeModifierKey
+            );
+
+            const baseInputType = getGqlInputTypeByTypeIdAndModifier(
+              baseTypeId,
+              typeModifierKey
+            );
+
+            gqlTypeByTypeIdAndModifier[type.id][typeModifierKey] = baseType;
             gqlInputTypeByTypeIdAndModifier[type.id][
               typeModifierKey
-            ] = Object.assign(Object.create(baseInputType), {
-              name: inflection.inputType(
-                gqlTypeByTypeIdAndModifier[type.id][typeModifierKey]
-              ),
+            ] = baseInputType;
+            shouldSkipAddType = true;
```

Comment:
This feels messy but I'm not sure how to solve it:

1. use `shouldSkipAddType` (messy)
2. return here directly (unexpected given rest of function does not)
3. have `addType` not throw in this circumstance (might hide programming error, though not sure what...?)

None of these are the clear winner but I think we should go with option 3 - let's make it so that `addType` does not throw if it receives the exact identical type again `(type1 === type2)`. Should be able to achieve this with a `Set` which is supported in Node 8 so we can use it: https://node.green/#ES2015-built-ins-Set

Reply:
I did 1 because I did not want to do 2 (the function is build as setting up a bunch of stuff, and then returning 1 thing, and I didn't want to violate that).

I considered 3, but decided not to do it because I recognized it as an error I found useful many times in development (if I tried to create a table `string` for example, I'd get a (good) error here about how String was already a registered type.

Is there a way to make `addType` not throw that doesn't also prevent the good things it catches?

Reply:
I think you may have misinterpreted my suggestion: addType would still throw on a name conflict, it just won’t throw if the exact type is already known; i.e. it becomes idempotent (`addType(addType(type))` would become allowed)

Reply:
Or maybe I misinterpreted your reply and you’re referring to the fallback String… in which case I can see how that would be useful. I’ve never faced this issue myself.

Reply:
Let's make `addType(addType(type))` idempotent but have it throw if you try and add any of the builtin scalars (`String`, `Int`, `Float`, `Boolean`, `ID`) or if you pass a `non-null` or `list` variant.

Reply:
I'm not exactly sure how to check if that exact type exists for the idempotency check (how to decide if this *exact type* is already known).

I'm seeing that each type has some combination of `name`, `description`, `astNode`, resolvers, etc.

Would a deep equality check on the `astNode` be the thing to do?

I'm still worried that we'd potentially be registering different resolvers for the same type sharing the same AST node, nd that could lead to unpredictable behavior.

Reply:
It turns out `addType` is already idempotent in this way?

https://github.com/graphile/graphile-engine/blob/726c210ae6f559c132f859fed18dc031a2a2cb1c/packages/graphile-build/src/makeNewBuild.js#L307

So this `shouldSkipAddType` shouldn't be necessary?